src/proxy-animation.js

```diff
@@ -30,151 +30,470 @@ function createReadyPromise(details) {
-    if (!internalPlayState)
-      return proxyAnimations.get(this).animation.startTime;
-    if (internalPlayState == "running")
-      return 0;
-    return null;
+    return getAnimationProperty(proxyAnimations.get(this), 'startTime');
   }
   set startTime(value) {
-    proxyAnimations.get(this).animation.startTime = value;
+    const previousCurrentTime = this.currentTime;
+    const details = proxyAnimations.get(this);
+
+    if (!details.timeline) {
+      details.animation.startTime = value;
+      return;
+    }
+
+    details.resetCurrentTimeOnResume = false;
+    applyPendingPlaybackRate(details);
+    details.readyPromise = null;
+    if (typeof value == 'number') {
+      details.holdTime = null;
+      details.startTime = value;
+      details.playState = 'running';
+      const timelineTime = details.timeline.currentTime;
+      details.animation.currentTime =
+          (timelineTime - details.startTime) * this.playbackRate;
+      updateFinishedState(details);
+    } else {
+      details.holdTime = previousCurrentTime;
+      details.startTime = null;
+      details.playState =
+          (typeof previousCurrentTime == 'number') ? 'paused' : 'idle';
+    }
   }
 
   get currentTime() {
-    return proxyAnimations.get(this).animation.currentTime;
+    const details = proxyAnimations.get(this);
+    if (details.timeline) {
+      if (!details.playState || details.playState == 'idle')
+        return null;
+      if (details.playState == 'running' &&
+          details.timeline.phase == 'inactive')
+        return null;
+    }
+    return details.animation.currentTime;
   }
   set currentTime(value) {
-    proxyAnimations.get(this).animation.currentTime = value;
+    const details = proxyAnimations.get(this);
+    details.animation.currentTime = value;
+    details.resetCurrentTimeOnResume = false;
+    if (details.timeline) {
+       // Update the start or the hold time of the proxy animation depending
+       // on the play sate.
+       const timelineTime = details.timeline.currentTime;
+       const playbackRate = this.playbackRate;
+       switch(details.playState) {
+        case 'running':
+        case 'finished':
+          // TODO: handle value == null or playbackRate == 0.
+          details.startTime = timelineTime - value / playbackRate;
+          details.holdTime = null;
+          break;
+
+        default:
+           details.playState = value ? 'paused' : 'idle';
+           details.holdTime = value;
+           details.startTime = null;
+           break;
+       }
+       updateFinishedState(details);
+    }
   }
 
   get playbackRate() {
     return proxyAnimations.get(this).animation.playbackRate;
   }
   set playbackRate(value) {
-    proxyAnimations.get(this).animation.playbackRate = value;
+    const details = proxyAnimations.get(this);
+    details.animation.playbackRate = value;
+    details.pendingPlaybackRate = null;
+    if (details.timeline)
```

Comment:
updateFinishedState is a noop if there is no timeline, we don't need this check.